### ISSUE TITLE
[Java] CWE-089: MyBatis Mapper XML SQL Injection

### ISSUE THREAD

**Issue by haby0** [github/securitylab#406]
## Query

*Link to pull request with your CodeQL query:*

Relevant PR: https://github.com/github/codeql/pull/6319

## CVE ID(s)

*List the CVE ID(s) associated with this vulnerability. GitHub will automatically link CVE IDs to the [GitHub Advisory Database](https://github.com/advisories).*

- [CVE-2019-3576](https://nvd.nist.gov/vuln/detail/CVE-2019-3576)

## Report

The MyBatis Mapper XML file allows the use of the $ character to construct dynamic SQL statements. Attackers can modify the meaning of statements or execute arbitrary SQL commands.

Note: CodeQL creates additional xml files for database analysis, you need to set the `LGTM_INDEX_XML_MODE` environment variable to `all`.

- [X] Are you planning to discuss this vulnerability submission publicly? (Blog Post, social networks, etc). *We would love to have you spread the word about the good work you are doing*

## Result(s)

*Provide at least one useful result found by your query, on some revision of a real project.*

- [source](https://github.com/201206030/novel-plus/blob/4540c3781e9bc1d47c388cb8b6058ed66e094586/novel-admin/src/main/java/com/java2nb/novel/controller/BookController.java#L49) -> [sink](https://github.com/201206030/novel-plus/blob/4540c3781e9bc1d47c388cb8b6058ed66e094586/novel-admin/src/main/resources/mybatis/novel/BookMapper.xml#L48)

**Comment by smowton** [comment]
I note some existing support for MyBatis in `MyBatis.qll`; could you describe how the two relate?

**Comment by haby0** [comment]
> I note some existing support for MyBatis in `MyBatis.qll`; could you describe how that relates to this submission?

`MyBatis.qll` This file is mainly for the `java` file that uses the `MyBatis` framework.

The file `MyBatisMapperXmlSqlInjectionLib.qll` in pr is mainly for the XML file of MyBatis Mapper.

I don't know if I have described it clearly.

**Comment by haby0** [comment]
CodeQL does not analyze all xml by default. I don't know how you can comprehensively evaluate this pr?

**Comment by antonio-morales** [comment]
Hi @haby0,

first of all, thanks for your submission.

In order to meet our new submission criteria, you need to support your submission with a CVE mapping (old or new). See https://securitylab.github.com/bounties for more details.

Would you be able to link this vulnerability with a CVE?

Regards,
Antonio.

**Comment by haby0** [comment]
> Hi @haby0,
> 
> first of all, thanks for your submission.
> 
> In order to meet our new submission criteria, you need to support your submission with a CVE mapping (old or new). See https://securitylab.github.com/bounties for more details.
> 
> Would you be able to link this vulnerability with a CVE?
> 
> Regards,
> Antonio.

No CVE yet.

**Comment by haby0** [comment]
@antonio-morales Hello, I found a historical cve (CVE-2019-3576) on the Internet, please refer to.

**Comment by haby0** [comment]
@antonio-morales Hello, I have improved the query configuration.

**Comment by haby0** [comment]
Friendly ping.

**Comment by antonio-morales** [comment]
Hi @haby0,

thanks for your reply.

I don't fully understand your submission, so I'm going to add my teammate @pwntester to this issue. Maybe he can give us his opinion on this.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **FP Check**.

_For information, the evaluation workflow is the following:
SecLab review > Generate Query Results > FP Check > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by pwntester** [comment]
Hi @haby0 Thanks for the submission, it looks great!

I found the open issue for the CVE you mentioned [here](https://gitee.com/inxeduopen/inxedu/issues/IQIIV)
And the code still looks [vulnerable](https://gitee.com/inxeduopen/inxedu/blob/master/demo_inxedu_open/src/main/resources/mybatis/inxedu/course/CourseFavoritesMapper.xml#L37)

Can you please check that project into GitHub and add it to LGTM so we can use it for verification purposes?

**Comment by pwntester** [comment]
I found a vulnerable version of the project affected by CVE-2019-3576 in https://github.com/Guangzhan/demo_inxedu_open/ I have added it to LGTM so it should be picked in the FP check run.

The author requires XML extraction:

> CodeQL creates additional xml files for database analysis, you need to set the LGTM_INDEX_XML_MODE environment variable to all

@smowton do you know if LGTM is already configured like that, otherwise the FP run is going to be pretty useless, right?

**Comment by pwntester** [comment]
Actually, scratch that! I found it on GitHub as well: https://github.com/Guangzhan/demo_inxedu_open/blob/master/src/main/resources/mybatis/inxedu/course/CourseFavoritesMapper.xml#L37 You can use it to verify your query finds the issue

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Generate Query Results**.

_For information, the evaluation workflow is the following:
SecLab review > Generate Query Results > FP Check > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by haby0** [comment]
@pwntester Please postpone it, I need to improve the query.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab review**.

_For information, the evaluation workflow is the following:
SecLab review > Generate Query Results > FP Check > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by haby0** [comment]
@pwntester Please review.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **FP Check**.

_For information, the evaluation workflow is the following:
SecLab review > Generate Query Results > FP Check > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by haby0** [comment]
@pwntester Hello, can you synchronize a progress now? Concerned about FP.

**Comment by haby0** [comment]
> @haby0 I think the CodeQL team was deciding what was the best way to extract MyBatis XML files on LGTM. Will ping them to see if there is any progress on that front

@pwntester The inclusion of the xml file when creating the database was mentioned in issues before.

We need to wait for them to solve this problem before we can perform FP Check?

**Comment by haby0** [comment]
Does LGTM now support extracting XML files by default?

Is it now complete to generate query results?

**Comment by smowton** [comment]
The relevant PR has landed, it should be live on lgtm.com in roughly 2-3 weeks

**Comment by smowton** [comment]
That support is now live; starting the query run.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Results analysis**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by haby0** [comment]
> That support is now live; starting the query run.

Is there already a result?

**Comment by pwntester** [comment]
> Is there already a result?

No, we will keep you posted when/if the issue moves to the next phase (Query review)

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Query review**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Pay**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by xcorail** [comment]
Created Hackerone report 1442954 for bounty 359691 : [406] [Java] CWE-089: MyBatis Mapper XML SQL Injection

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Closed**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_
